Fall back to EGL for GL replay

afrantzis opened the pull request:
## Description

We want to try to use the EGL platform as a fallback when the GL
platform can't create a GL context, since EGL_DEFAULT_DISPLAY might
still provide a workable EGLDisplay. This, for example, allows us to
replay traces when on-screen display is not required, by using the Mesa
surfaceless platform (by setting the "EGL_PLATFORM=surfaceless" Mesa
environment variable).

If we prefer to not have such an implicit fallback, an alternative would be to
provide an explicit option (e.g., in ReplayOptions or in an environment
variable?) to force a particular platform.

afrantzis committed:
Fall back to EGL for GL replay

We want to try to use the EGL platform as a fallback when the GL
platform can't create a GL context, since EGL_DEFAULT_DISPLAY might
still provide a workable EGLDisplay. This, for example, allows us to
replay traces when on-screen display is not required, by using the Mesa
surfaceless platform (by setting the "EGL_PLATFORM=surfaceless" Mesa
environment variable).

baldurk commented:
Can you clarify what API you are using? Are you using GLES or OpenGL proper? This PR is modifying the OpenGL proper replay code, but then pulling in a dependency on the GLES support. It makes for more confusing logic and it doesn't handle the case where GLES support is not compiled in (does that even make sense when you are wanting to use GLES's windowing API as a fallback? I don't know).

I'm not opposed to having a fallback necessarily, but I'd like to understand more clearly the motivations. I don't quite understand what good it does to be able to load a replay if you don't have a display, since then you then can't actually do anything interesting with the capture.

afrantzis commented:
Hi! The use case that inspired this change is being able to replay traces in a CI environment (a [real-world example](https://gitlab.freedesktop.org/gfx-ci/tracie/tracie)) and dump the resulting images, in a way that doesn't involve any on-screen display or other interaction with the windowing system. At the moment, in order to replay/dump images, we need to set up an xserver (e.g., with xvfb), but this is problematic since it introduces another component which we don't necessarily want to test (window system integration) and generally complicates things in CI. Plus upstream xvfb doesn't currently support hardware rendering (although we are working on this, too). Using the Mesa "surfaceless" platform solves this issue more elegantly, since we can just use EGL_DEFAULT_DISPLAY to get a functional EGLDisplay without any windowing system integration. 

For the replays, both desktop GL and GLES are of interest. For the GL case, I agree that the GL path falling back to the GLES path just to get EGL for GL is somewhat confusing, but that was preexisting for Wayland support (note that Wayland can work with just GL, too) and in this PR I piggybacked on this approach. To make things clearer and remove the dependency on GLES, I think that in `gl_replay.cpp` we could replace the fallback call to `GLES_CreateReplayDevice()` with something like the following in `GL_CreateReplayDevice()`:

```c++
bool load_ok = GetEGLPlatform().PopulateForReplay();

if(!load_ok)
{
  RDCERR("Couldn't find required EGL function addresses");
  return ReplayStatus::APIInitFailed;
}

RDCLOG("Initialising GL replay via libEGL");

return CreateReplayDevice(rdc ? rdc->GetDriver() : RDCDriver::OpenGL, rdc, opts,
                          GetEGLPlatform(), driver);
```

If I am not missing some adverse effect this would have on other users of this code, and this sounds interesting, I would be happy to propose something along these lines.

baldurk commented:
OK fair enough, it's annoyingly ugly that GL is starting to grow dependencies on GLES's WSI API but I can see how GLX being awful in its intrinsic link to the display causes problems too. The wayland stuff should pretty much be ignored - wayland is currently not supported on replay and is pretty broken, the code there is only as a first pass.

I think then the solution is to take EGL out of being GLES-conditional and compile it in when either GL or GLES is supported. That will mean breaking any dependency EGL has on GLES, I don't know if what the headers/function calls do.

Then we can add a parallel equivalent into `GLPlatform`: `CanCreateGLContext` which is similar to `CanCreateGLESContext`. For `WGLPlatform`/`CGLPlatform` etc this would return trivially true, for `GLXPlatform` it would check to see if an xlib display is available, and for EGL it could check to see if GL is available via GL (is there a query for whether the hacky GL-over-EGL is available? If not is it safe to assume it's always available on linux? It should at least return false on Android).

In that way the `GL_CreateReplayDevice` can then have a fallback path - first it tries to use GLPlatform(), but if that claims it cannot create a GL context then it falls back and tries to use EGL. Much the equivalent of 

The wayland hack to force use of EGL should remain separate to keep things simpler, until such a time as wayland is properly supported and it can then be integrated somewhere into the above logic.

afrantzis commented:
Hi! Here is my take on the above: https://github.com/baldurk/renderdoc/pull/1549 .
Two notes, slight differences from what was discussed:
* I have made EGL support independently configurable at build time, to allow builds without EGL at all if required.
* PR #1549 doesn't implement any EGL fallback for GL, just keeps the wayland special case. I'll update this PR (#1546) to implement the fallback once we are happy with the other changes.

afrantzis commented:
Hi! I rebased this proposal on the latest v1.x, and updated it to use a fallback path for EGL/GL while leaving the wayland path independent, as discussed above.
